web/packages/teleport/src/Locks/useGetTargetData.tsx

```diff
@@ -64,30 +65,42 @@ export const useGetTargetData: UseGetTargetData = (
       user: {
         fetch: fetchUsers,
         handler: (setter, users) => {
-          const filteredData = users.map(u => ({
-            name: u.name,
-            roles: u.roles.join(', '),
-          }));
+          const filteredData = users.map(
+            u =>
+              ({
+                name: u.name,
+                roles: u.roles.join(', '),
+                targetValue: u.name,
+              } as TableData)
```

Comment:
I am not a fan of these `as TableData` casts, but the implicit any typing all over the place made it hard for me to enforce that `targetValue` is provided.

Would love alternate suggestions!

Reply:
Slightly more elegant alternative is to add a type to a variable: `const filteredData: TableData[]`.
But in general, I feel like this structure with `fetch`, `handler` (and especially `setter` that could just return the value) and `options` is unnecessary complicated. We split a single request into an object with a few properties that are later combined back to a single request. This makes it quite hard to make all the properties type safe.

I'd see the structure more or less this way:

```ts
import {useMemo} from "react";
import {AllowedTargets, TableData} from "./types";

  const targetDataFilters = useMemo<
    Partial<Record<AllowedTargets, { fetchData(): Promise<TableData[]> }>>
  >(() => {
    return {
      user: {
        fetchData: async () => {
          const users = await fetchUsers();
          return users.map(
            u =>
            ({
              name: u.name,
              roles: u.roles.join(', '),
              targetValue: u.name,
            })
          );
        },
      },
      role: {
        fetchData: async () => {
          const roles = await fetchRoles();
          return roles.map(
            r =>
            ({
              name: r.name,
              targetValue: r.name,
            })
          );
        },
      },
      node: {
        fetchData: async () => {
          const nodes = await fetchNodes(clusterId, { limit: 10 });
          return nodes.agents.map(
            n =>
            ({
              name: n.hostname,
              addr: n.addr,
              labels: n.labels,
              targetValue: n.id,
            })
          );
        },
      },
      mfa_device: {
        fetchData: async () => {
          const mfas = await fetchDevices();
          return mfas.map(
            m =>
            ({
              name: m.name,
              id: m.id,
              description: m.description,
              lastUsed: m.lastUsedDate.toUTCString(),
              targetValue: m.id,
            })
          );
        },
      },
      windows_desktop: {
        fetchData: async () => {
          const desktops = await fetchDesktops(clusterId, { limit: 10 });
          return desktops.agents.map(
            d =>
            ({
              name: d.name,
              addr: d.addr,
              labels: d.labels,
              targetValue: d.name,
            })
          );
        },
      },
    };
  }, [
    clusterId,
    fetchDesktops,
    fetchDevices,
    fetchNodes,
    fetchRoles,
    fetchUsers,
  ]);
```
But as Lisa wrote, this requires some refactoring - also for `additionalLocks` that shares the same structure. 

I also noticed that "Create New Lock" screen doesn't have any loading and error states (errors are silently ignored) :/